File: AllInOne/src/DyanmicProgramming/WildcardMatching.java
package DyanmicProgramming;

import java.util.Arrays;

/**
 * 
 * @author jingjiejiang Jan 22, 2019
 * @history Aug 30, 2020
 *
 */
public class WildcardMatching {

    // DP:two demensional
	public boolean isMatch(String s, String p) {
        
        if (s == null) return false;
        
        boolean[][] dp = new boolean[p.length() + 1][s.length() + 1];
        dp[0][0] = true;
        
        for (int row = 1; row <= p.length(); row ++) {
            dp[row][0] = p.charAt(row - 1) == '*'? dp[row - 1][0] : false; 
        }
        
        for (int row = 1; row <= p.length(); row ++) {
            
            char pChar = p.charAt(row - 1);
            for (int col = 1; col <= s.length(); col ++) {
                
                char sChar = s.charAt(col - 1);
                
                if (sChar == pChar || pChar == '?') {
                    dp[row][col] = dp[row - 1][col - 1];
                } else if (pChar == '*') {
                    dp[row][col] = dp[row - 1][col - 1] || dp[row][col - 1] || dp[row - 1][col];
                } else {
                    dp[row][col] = false;
                }
            }
        }

        return dp[p.length()][s.length()];
    }
}

File: AllInOne/src/DyanmicProgramming/WildcardMatching.js
/**
 * 
 * 44. Wildcard Matching
 * 
 */
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
 var isMatch = function(s, p) {
    
  let dp = Array(p.length + 1).fill(null).map(_ => Array(s.length + 1).fill(false));
    
  dp[0][0] = true;
  
  for (let row = 1; row <= p.length; row ++) {
    dp[row][0] = (p[row - 1] === '*' ? dp[row - 1][0] : false); 
  }

  for (let row = 1; row <= p.length; row ++) {
    for (let col = 1; col <= s.length; col ++) {

      if (p[row - 1] === s[col - 1] || p[row - 1] === '?') {
        dp[row][col] = dp[row - 1][col - 1];
      } else if (p[row - 1] === '*') {
        dp[row][col] = dp[row - 1][col - 1] || dp[row - 1][col] || dp[row][col - 1];
      } else {
        dp[row][col] = false;
      }
    }
  }

  return dp[p.length][s.length];
};